File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/emit.c -- C-side emitter for the rocke parity harness.
 *
 * Builds one of four kernels (selected by argv[1]) identically to the Python
 * emitter in emit.py and prints the lowered AMDGPU LLVM .ll to stdout, so the
 * two outputs can be byte-compared.
 *
 *   scalar  : c = const(1); r = c + c
 *   memory  : params -> tid -> 2x global_load_f32 -> fadd -> global_store
 *   forloop : scf.for accumulating loop (exercises scf.for port)
 *   vector  : vector splat + vector binop + vector fptrunc/extract
 *
 * arch = gfx950, flavor = AUTO (matches Python defaults in this harness).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/ir.h"
#include "rocke/lower_llvm.h"

static int build_scalar(rocke_ir_builder_t* b)
{
    rocke_value_t* c = rocke_b_const_i32(b, 1);
    rocke_value_t* r = rocke_b_add(b, c, c);
    (void)r;
    rocke_b_ret(b);
    return 0;
}

static int build_memory(rocke_ir_builder_t* b)
{
    const rocke_type_t* pf32 = rocke_ptr_type(b, rocke_f32(), "global");
    rocke_value_t* A = rocke_b_param(b, "A", pf32, NULL);
    rocke_value_t* B = rocke_b_param(b, "B", pf32, NULL);
    rocke_value_t* C = rocke_b_param(b, "C", pf32, NULL);
    rocke_value_t* tid = rocke_b_thread_id_x(b);
    rocke_value_t* a = rocke_b_global_load_f32(b, A, tid, 4);
    rocke_value_t* bb = rocke_b_global_load_f32(b, B, tid, 4);
    rocke_value_t* s = rocke_b_fadd(b, a, bb);
    rocke_b_global_store(b, C, tid, s, 4);
    rocke_b_ret(b);
    return 0;
}

static int build_forloop(rocke_ir_builder_t* b)
{
    const rocke_type_t* pf32 = rocke_ptr_type(b, rocke_f32(), "global");
    rocke_value_t* C = rocke_b_param(b, "C", pf32, NULL);
    rocke_value_t* lo = rocke_b_const_i32(b, 0);
    rocke_value_t* hi = rocke_b_const_i32(b, 16);
    rocke_value_t* step = rocke_b_const_i32(b, 1);
    rocke_value_t* acc0 = rocke_b_const_f32(b, 0.0);
    rocke_iter_arg_t iters[1];
    iters[0].name = "acc";
    iters[0].init = acc0;
    rocke_for_t f = rocke_b_scf_for_iter(b,
                                         lo,
                                         hi,
                                         step,
                                         iters,
                                         1,
                                         "k0",
                                         /*unroll=*/false,
                                         /*elide_trailing_barrier=*/true);
    rocke_b_region_enter(b, f.body);
    {
        rocke_value_t* acc = f.iter_vars[0];
        rocke_value_t* one = rocke_b_const_f32(b, 1.0);
        rocke_value_t* nacc = rocke_b_fadd(b, acc, one);
        rocke_value_t* yld[1];
        yld[0] = nacc;
        rocke_b_scf_yield(b, yld, 1);
    }
    rocke_b_region_leave(b);
    rocke_value_t* tid = rocke_b_thread_id_x(b);
    rocke_b_global_store(b, C, tid, f.op->results[0], 4);
    rocke_b_ret(b);
    return 0;
}

static int build_vector(rocke_ir_builder_t* b)
{
    const rocke_type_t* pf16 = rocke_ptr_type(b, rocke_f16(), "global");
    rocke_value_t* C = rocke_b_param(b, "C", pf16, NULL);
    rocke_value_t* s = rocke_b_const_f32(b, 2.0);
    rocke_value_t* v = rocke_b_vector_splat(b, s, 4); /* <4 x f32> */
    rocke_value_t* w = rocke_b_vector_add(b, v, v); /* <4 x f32> */
    rocke_value_t* h = rocke_b_vec_trunc_f32_to_f16(b, w); /* <4 x f16> */
    rocke_value_t* e = rocke_b_vec_extract(b, h, 0); /* f16 */
    rocke_value_t* tid = rocke_b_thread_id_x(b);
    rocke_b_store_f16(b, C, tid, e);
    rocke_b_ret(b);
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <scalar|memory|forloop|vector>\n", argv[0]);
        return 2;
    }
    const char* which = argv[1];

    rocke_ir_builder_t b;
    if(rocke_ir_builder_init(&b, "parity_kernel") != ROCKE_OK)
    {
        fprintf(stderr, "builder init failed\n");
        return 1;
    }

    int rc;
    if(strcmp(which, "scalar") == 0)
        rc = build_scalar(&b);
    else if(strcmp(which, "memory") == 0)
        rc = build_memory(&b);
    else if(strcmp(which, "forloop") == 0)
        rc = build_forloop(&b);
    else if(strcmp(which, "vector") == 0)
        rc = build_vector(&b);
    else
    {
        fprintf(stderr, "unknown kernel %s\n", which);
        rocke_ir_builder_free(&b);
        return 2;
    }
    (void)rc;

    if(!rocke_ir_builder_ok(&b))
    {
        fprintf(stderr, "builder error: %s\n", rocke_ir_builder_error(&b));
        rocke_ir_builder_free(&b);
        return 1;
    }

    rocke_kernel_def_t* kernel = rocke_ir_builder_kernel(&b);
    char* llvm_text = NULL;
    char err[ROCKE_ERR_MSG_CAP];
    err[0] = 0;
    rocke_status_t st = rocke_lower_kernel_to_llvm_ex(
        kernel, ROCKE_LLVM_FLAVOR_AUTO, "gfx950", &llvm_text, err, sizeof err);
    if(st != ROCKE_OK || !llvm_text)
    {
        fprintf(stderr, "lower failed: status=%d err=%s\n", (int)st, err);
        rocke_ir_builder_free(&b);
        return 1;
    }
    fputs(llvm_text, stdout);
    free(llvm_text);
    rocke_ir_builder_free(&b);
    return 0;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/emit.py -- Python-side emitter for the rocke parity harness.
#
# Builds one of four kernels (selected by argv[1]) identically to the C emitter
# in emit.c and prints the lowered AMDGPU LLVM .ll to stdout, so the two outputs
# can be byte-compared.
#
#   scalar  : c = const(1); r = c + c
#   memory  : params -> tid -> 2x global_load_f32 -> fadd -> global_store
#   forloop : scf.for accumulating loop (exercises scf.for port)
#   vector  : vector splat + vector binop + vector fptrunc/extract
#
# arch = gfx950, llvm_flavor = None (AUTO), matching the C side.
import sys

from rocke.core.ir import IRBuilder, F16, F32, PtrType
from rocke.core.lower_llvm import lower_kernel_to_llvm


def build_scalar(b: IRBuilder) -> None:
    c = b.const_i32(1)
    _ = b.add(c, c)
    b.ret()


def build_memory(b: IRBuilder) -> None:
    pf32 = PtrType(F32, "global")
    A = b.param("A", pf32)
    B = b.param("B", pf32)
    C = b.param("C", pf32)
    tid = b.thread_id_x()
    a = b.global_load_f32(A, tid, align=4)
    bb = b.global_load_f32(B, tid, align=4)
    s = b.fadd(a, bb)
    b.global_store(C, tid, s, align=4)
    b.ret()


def build_forloop(b: IRBuilder) -> None:
    pf32 = PtrType(F32, "global")
    C = b.param("C", pf32)
    lo = b.const_i32(0)
    hi = b.const_i32(16)
    step = b.const_i32(1)
    acc0 = b.const_f32(0.0)
    loop = b.scf_for_iter(
        lo,
        hi,
        step,
        [("acc", acc0)],
        iv_name="k0",
        unroll=False,
        elide_trailing_barrier=True,
    )
    with loop as (k0, iters):
        acc = iters[0]
        one = b.const_f32(1.0)
        nacc = b.fadd(acc, one)
        b.scf_yield(nacc)
    tid = b.thread_id_x()
    b.global_store(C, tid, loop.results[0], align=4)
    b.ret()


def build_vector(b: IRBuilder) -> None:
    pf16 = PtrType(F16, "global")
    C = b.param("C", pf16)
    s = b.const_f32(2.0)
    v = b.vector_splat(s, 4)  # <4 x f32>
    w = b.vector_add(v, v)  # <4 x f32>
    h = b.vec_trunc_f32_to_f16(w)  # <4 x f16>
    e = b.vec_extract(h, 0)  # f16
    tid = b.thread_id_x()
    b.store_f16(C, tid, e)
    b.ret()


BUILDERS = {
    "scalar": build_scalar,
    "memory": build_memory,
    "forloop": build_forloop,
    "vector": build_vector,
}


def main() -> int:
    if len(sys.argv) < 2 or sys.argv[1] not in BUILDERS:
        sys.stderr.write("usage: emit.py <scalar|memory|forloop|vector>\n")
        return 2
    which = sys.argv[1]
    b = IRBuilder("parity_kernel")
    BUILDERS[which](b)
    text = lower_kernel_to_llvm(b.kernel, llvm_flavor=None, arch="gfx950")
    sys.stdout.write(text)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
